src/nxt_http_return.c

```diff
@@ -87,27 +97,123 @@ nxt_http_return(nxt_task_t *task, nxt_http_request_t *r,
         return NULL;
     }
 
+    ctx = nxt_mp_zget(r->mem_pool, sizeof(*ctx));
```

Comment:
We usually use structure type.
```
ctx = nxt_mp_zget(r->mem_pool, sizeof(nxt_http_return_ctx_t));
```

Reply:
Hmm yes.  That's something I don't really like.

I think it's much safer to use the pointer itself to get the size.  That way you can never make a mistake, since it's straightforward (to the point that I have a macro that does it for me: `err = alx_mallocarrays(&p, n);` which basically resolves to `p = malloc(sizeof(*p) * n); err = (p == NULL);` plus some overflow checking).  However, if you use the type, and you ever change the type of the pointer, you need to update _every_ allocation line.

However, for consistency, I'll use the type here.  I'll propose in a different thread to revise the allocations.

<https://software.codidact.com/posts/285898/285899#answer-285899>
<https://stackoverflow.com/a/605858/6872717>

Reply:
@alejandro-colomar If you decide to change the type of pointer, then you anyway need to check every line for possible consequences. It's a must have habit no matter of how you allocate memory.

Note also, that missing `*` in such type of allocation is [a very common mistake](https://cwe.mitre.org/data/definitions/467.html). And the argument that if you use pointer, then you will never allocate wrong size won't work at the moment when you decide to change to an array type, and then having no habit to check allocation may play trick to you.

The main reason to use the type name in allocations is that number of times when someone has to find a place where a particular object is allocated by several orders of magnitude bigger than changes of its type. Currently it's easy to find where particular objects are created by `alloc.*nxt_some_type_t` lookup request. Programmers read the code much more often than change it, and having a type in alocations makes searching and reading a bit easier.

Reply:
> @alejandro-colomar If you decide to change the type of pointer, then you anyway need to check every line for possible consequences. It's a must have habit no matter of how you allocate memory.

Yes, but it's easier to make mistakes in the case of a type.

> 
> Note also, that missing `*` in such type of allocation is [a very common mistake](https://cwe.mitre.org/data/definitions/467.html). And the argument that if you use pointer, then you will never allocate wrong size won't work at the moment when you decide to change to an array type, and then having no habit to check allocation may play trick to you.

I have the rule that the argument to `sizeof()` must the be exact same string to the left of the `=` with a prefix of exactly one `*`.  Since it's a text replacement rule, I created a macro that does it for me.  I (almost) always allocate arrays, of size 1 when I want a single element:

`err = alx_mallocarrays(&ctx, 1);`

A trivial definition would be:

```
#define alx_mallocarray(nmemb, size)  reallocarray(NULL, nmemb, size)

#define alx_mallocarrays(ptr, nmemb)                 \
({                                                   \
    int          err_;                               \
    __auto_type  ptr_ = (ptr);                       \
                                                     \
    *ptr_ = alx_mallocarray(nmemb, sizeof(**ptr_));  \
    err_ = (*ptr == NULL);                           \
                                                     \
    err_;                                            \
})
```

Only when I want to specify a size that is not a multiple of the type I do call malloc directly, for example when I need to use the struct hack.

This has always worked for me.

So regarding safety, I believe this macro is much safer than any kind of `sizeof` usage.  And lacking this macro, a mental rule that always works (i.e., doing mentally the same work that this macro does) is probably the safest, but not perfect.

> 
> The main reason to use the type name in allocations is that number of times when someone has to find a place where a particular object is allocated by several orders of magnitude bigger than changes of its type. Currently it's easy to find where particular objects are created by `alloc.*nxt_some_type_t` lookup request. Programmers read the code much more often than change it, and having a type in alocations makes searching and reading a bit easier.

However, that is a strong argument.  I like being able to grep with simple patterns such as that one.  And it's true that we grep much more than we edit.